### PR TITLE
feat: implement fit() pipeline verb

### PR DESCRIPTION
**Pull request by S-Leuthold**
## Summary

Implements `fit()`, the fourth pipeline verb in the horizons workflow chain (`spectra → configure → validate → evaluate → fit`). This verb re-tunes the top N configurations from `evaluate()` using warm-start Bayesian optimization, generates out-of-fold CV predictions for downstream stacking, fits final deployable models on held-out test data, and optionally trains CQR-based uncertainty quantification components.

- **Capture layer** (`fit-single-config.R`): 15-step inner loop that runs silently and returns structured results — warm-start tuning, CV predictions, final fit, back-transformation, degradation detection, UQ, butchering
- **Warm-start engine** (`fit-warmstart.R`): Builds exploration grids centered on evaluate's best params, runs grid search + optional Bayesian refinement
- **UQ system** (`fit-uq.R`): Conformalized Quantile Regression — trains quantile ranger on OOF residuals, calibrates with conformal scores on held-out calibration set
- **Render layer** (`pipeline-fit.R`): Tree CLI output matching evaluate's patterns — preflight validation, data partitioning, config loop with test/CV comparison, degradation flags, UQ coverage
- **Print/summary methods**: Class-aware display showing evaluation results and models info at each pipeline stage
- **Integration tests** (`test-pipeline-fit.R`): 37 tests covering preflight, success path, CV predictions, results tibble, row index, UQ enabled/disabled, edge cases, seed reproducibility

### New files
| File | LOC | Purpose |
|------|-----|---------|
| `R/pipeline-fit.R` | 595 | Render layer (main verb) |
| `R/fit-single-config.R` | 535 | Capture layer (inner loop) |
| `R/fit-warmstart.R` | 325 | Warm-start grid + Bayesian tuning |
| `R/fit-uq.R` | 284 | CQR uncertainty quantification |
| `tests/testthat/test-pipeline-fit.R` | 538 | Integration tests |
| `tests/testthat/test-fit-single-config.R` | 609 | Capture layer tests |
| `tests/testthat/test-fit-warmstart.R` | 425 | Warm-start tests |
| `tests/testthat/test-fit-uq.R` | 322 | UQ tests |

## Test plan

- [x] `devtools::test(filter = "pipeline-fit")` — 37 tests pass
- [x] `devtools::test(filter = "fit-single-config")` — 67 tests pass  
- [x] `devtools::test(filter = "fit-warmstart")` — 52 tests pass
- [x] `devtools::test(filter = "fit-uq")` — 33 tests pass
- [x] Full suite: 972 tests pass, 0 failures, 7 pre-existing warnings
- [x] Print/summary smoke test across all three class levels

## Summary by CodeRabbit

* **New Features**
  * Warm-start grid + Bayesian re-tuning pipeline for best configs; single-config capture/retrain flow exposed.
  * Optional conformalized uncertainty quantification to produce calibrated prediction intervals.

* **Improvements**
  * Richer printed summaries and summaries with evaluation/models sections, metrics, timestamps, and runtime.
  * Safer, more reproducible fitting with fallbacks, degradation detection, and per-config status reporting.

* **Documentation**
  * Docs added/updated for new APIs and internal helpers.

* **Tests**
  * Extensive test coverage for tuning, fitting, and UQ workflows.

### CODE REVIEW

File: man/compute_c_alpha.Rd [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+\keyword{finite-sample}
+\keyword{internal}
+\keyword{level,}
+\keyword{nonconformity}
+\keyword{of}
+\keyword{scores}
+\keyword{standard}
+\keyword{the}
+\keyword{using}
+\keyword{vector}
```
[the remark this answers is not in the document]
Comment (S-Leuthold) [reply man/compute_c_alpha.Rd]: ✅ Fixed — removed orphaned roxygen block that caused keyword spam. Regenerated Rd.

File: man/fit_single_config.Rd [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+Inner-loop function for \code{fit()}. Takes a single config row from the
+evaluate() leaderboard, re-tunes it with warm-start Bayesian optimization,
+generates OOF predictions, fits a final deployable model, evaluates on
+the held-out test set, and detects performance degradation.
+
+This function is the \strong{capture layer} — it runs silently and returns a
+structured list. All console output (tree rendering, progress messages) is
+the responsibility of the calling \code{fit()} function.
+}
+\keyword{internal}
```
[the remark this answers is not in the document]
Comment (S-Leuthold) [reply man/fit_single_config.Rd:64]: ✅ Acknowledged — @keywords internal + @export is intentional (exported but hidden from help index). This is standard R package practice for power-user functions.

File: man/fit_uq.Rd [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+Trains a quantile random forest on OOF residuals from the point model,
+then computes conformal nonconformity scores on the held-out calibration
+data. Returns everything needed for \code{predict()} to produce prediction
+intervals at any coverage level.
+}
+\details{
+This function is the \strong{capture layer} — it runs silently and returns a
+structured list. Messaging is the caller's responsibility.
+}
+\keyword{internal}
```
[the remark this answers is not in the document]
Comment (S-Leuthold) [reply man/fit_uq.Rd:47]: ✅ Acknowledged — same pattern as Issue 2. @keywords internal + @export is deliberate.

File: R/class-core.R [hunk 4]
```diff
@@ -570,6 +581,101 @@ print.horizons_data <- function(x, ...) {
...
+
+    eval_res  <- x$evaluation$results
+    n_success <- sum(eval_res$status == "success")
+    n_total   <- nrow(eval_res)
+
+    has_models <- !is.null(x$models$workflows)
+
+    cat(paste0("   \u251C\u2500 Configs evaluated: ", n_total, "\n"))
+    cat(paste0("   \u251C\u2500 Successful: ", n_success, "\n"))
+
```
[the remark this answers is not in the document]
Comment (S-Leuthold) [reply R/class-core.R:602]: ✅ Fixed — added na.rm = TRUE to all status count calculations in print().

File: R/class-core.R [hunk 5]
```diff
@@ -844,6 +960,180 @@ summary.horizons_data <- function(object, ...) {
...
+      if (nrow(best_row) > 0) {
+
+        ## Look up model display name from configs
+        cfg_row    <- x$config$configs[x$config$configs$config_id == best_id, ]
+        model_name <- if (nrow(cfg_row) > 0) {
+          MODEL_DISPLAY_NAMES[cfg_row$model] %||% cfg_row$model
+        } else {
+          best_id
+        }
```
[the remark this answers is not in the document]
Comment (S-Leuthold) [reply R/class-core.R:1003]: ✅ Fixed — MODEL_DISPLAY_NAMES now checks is.na() explicitly instead of relying on %||% (which does not catch NA).

File: R/class-core.R [hunk 6]
```diff
@@ -844,6 +960,180 @@ summary.horizons_data <- function(object, ...) {
...
+
+      n_success  <- sum(fit_res$status == "success")
+      n_degraded <- sum(fit_res$degraded, na.rm = TRUE)
+      n_failed   <- sum(fit_res$status == "failed")
+
+      cat(paste0("   \u251C\u2500 Results: ",
+                 cli::col_green(paste0(n_success, " success")),
+                 if (n_degraded > 0) paste0(" (", cli::col_yellow(paste0(n_degraded, " degraded")), ")") else "",
+                 if (n_failed > 0) paste0(", ", cli::col_red(paste0(n_failed, " failed"))) else "",
+                 "\n"))
```
[the remark this answers is not in the document]
Comment (S-Leuthold) [reply R/class-core.R]: ✅ Fixed — added na.rm = TRUE to n_success and n_failed counts in summary().

File: R/constants.R [hunk 7]
```diff
@@ -107,6 +107,14 @@ KNOWN_SPATIAL_COVARIATES <- c(
   "hillshade", "roughness"
 )
 
+## fit() Parameters ------------------------------------------------------------
+
+DEFAULT_FINAL_BAYES_ITER <- 25
+N_CALIB_MIN              <- 30L
+DEFAULT_UQ_LEVEL         <- 0.90
+UQ_QUANTILE_TREES        <- 500L
+WARMSTART_GRID_SIZE      <- 25L
```
[the remark this answers is not in the document]
Comment (S-Leuthold) [reply R/constants.R]: ✅ Fixed — changed DEFAULT_FINAL_BAYES_ITER from 25 to 25L.

File: R/fit-single-config.R [hunk 8]
```diff
@@ -0,0 +1,535 @@
...
+                              seed                = 42L) {
+
+  start_time <- Sys.time()
+  set.seed(seed)
+
+  config_id      <- config_row$config_id
+  outcome_col    <- role_map$variable[role_map$role == "outcome"]
+  transformation <- tolower(as.character(config_row$transformation))
+  train_data     <- rsample::training(split_F)
+  test_data      <- rsample::testing(split_F)
```
[the remark this answers is not in the document]
Comment (S-Leuthold) [reply R/fit-single-config.R:54]: ✅ Fixed — added train_data parameter to fit_single_config (defaults to training(split_F)). pipeline-fit.R now passes train_Fit to avoid UQ calibration leakage.

File: R/fit-single-config.R [hunk 9]
```diff
@@ -0,0 +1,535 @@
...
+  tune_result <- safely_execute(
+    tune_warmstart_bayes(
+      workflow      = wflow,
+      cv_resamples  = cv_resamples,
+      best_params   = best_params_eval,
+      param_set     = param_set,
+      bayesian_iter = final_bayesian_iter,
+      metric_set    = tune_metrics,
+      allow_par     = allow_par
+    ),
```
[the remark this answers is not in the document]
Comment (S-Leuthold) [reply R/fit-single-config.R:218]: ✅ Fixed — grid_size is now forwarded from fit_single_config to tune_warmstart_bayes.

File: R/fit-uq.R [hunk 10]
```diff
@@ -0,0 +1,284 @@
...
+compute_c_alpha <- function(scores, level) {
+
+  alpha  <- 1 - level
+  n      <- length(scores)
+
+  ## Finite-sample correction: ceiling to be conservative
+  q_prob <- min(1, ceiling((1 - alpha) * (n + 1)) / n)
+
+  stats::quantile(scores, probs = q_prob, type = 7, names = FALSE)
+
```
[the remark this answers is not in the document]
Comment (S-Leuthold) [reply R/fit-uq.R:46]: ✅ Fixed — compute_c_alpha now validates level is numeric in (0,1), filters non-finite scores, and aborts on empty score vector. Updated tests to cover boundary cases.

File: R/fit-uq.R [hunk 11]
```diff
@@ -0,0 +1,284 @@
...
+  outcome_col <- role_map$variable[role_map$role == "outcome"]
+
+  ## --- Guard: minimum calibration size -------------------------------------
+
+  if (nrow(calib_data) < N_CALIB_MIN) {
+
+    return(NULL)
+
+  }
```
[the remark this answers is not in the document]
Comment (S-Leuthold) [reply R/fit-uq.R:95]: ✅ Fixed — fit_uq now checks is.null(calib_data) before calling nrow().

File: R/fit-warmstart.R [hunk 12]
```diff
@@ -0,0 +1,325 @@
...
+    multipliers <- c(0.8, 0.9, 1.0, 1.1, 1.2)
+    candidates <- best_val * multipliers
+
+    candidates <- pmax(candidates, lower)
+    candidates <- pmin(candidates, upper)
+
+  }
+
+  unique(candidates)
+
```
[the remark this answers is not in the document]
Comment (S-Leuthold) [reply R/fit-warmstart.R:203]: ✅ Fixed — generate_candidates now handles NA best_val by generating evenly-spaced candidates across the full parameter range.

File: R/fit-warmstart.R [hunk 13]
```diff
@@ -0,0 +1,325 @@
...
+                                  best_params,
+                                  param_set,
+                                  bayesian_iter,
+                                  metric_set,
+                                  allow_par = FALSE) {
+
+  ## --- Phase 1: Build warm-start grid and run tune_grid -------------------
+
+  initial_grid <- build_warmstart_grid(best_params, param_set)
+  fallback_used <- !usable_best_params(best_params, param_set)
```
[the remark this answers is not in the document]
Comment (S-Leuthold) [reply R/fit-warmstart.R:276]: ✅ Fixed — tune_warmstart_bayes now accepts grid_size parameter and forwards it to build_warmstart_grid.

File: R/pipeline-fit.R [hunk 14]
```diff
@@ -0,0 +1,595 @@
...
+
+    rank_order <- order(metric_vals, decreasing = TRUE)
+
+  } else {
+
+    rank_order <- order(metric_vals, decreasing = FALSE)
+
+  }
+
+  successes <- successes[rank_order, ]
```
[the remark this answers is not in the document]
Comment (S-Leuthold) [reply R/pipeline-fit.R:105]: ✅ Fixed — pipeline-fit now validates rank_metric exists in successes and aborts if all values are NA.

File: R/pipeline-fit.R [hunk 15]
```diff
@@ -0,0 +1,595 @@
...
+      cv_resamples        = cv_resamples,
+      calib_data          = calib_data,
+      role_map            = role_map,
+      best_params_eval    = best_params_eval,
+      final_bayesian_iter = tuning$bayesian_iter,
+      grid_size           = tuning$grid_size,
+      compute_uq          = compute_uq,
+      allow_par           = allow_par,
+      seed                = seed
+    )
```
[the remark this answers is not in the document]
Comment (S-Leuthold) [reply R/pipeline-fit.R:330]: ✅ Fixed — pipeline-fit now passes train_Fit explicitly to fit_single_config, preventing calibration rows from leaking into the final model when compute_uq=TRUE.

File: R/pipeline-fit.R [hunk 16]
```diff
@@ -0,0 +1,595 @@
...
+      runtime_secs    = res$runtime_secs
+    )
+
+  })
+
+  ## Build row_index: .row → sample_id mapping from train_Fit
+  row_index <- tibble::tibble(
+    .row      = seq_len(nrow(train_Fit)),
+    sample_id = train_Fit[["sample_id"]]
+  )
```
[the remark this answers is not in the document]
Comment (S-Leuthold) [reply R/pipeline-fit.R]: ✅ Fixed — id column is now derived from role_map instead of hard-coding sample_id. Falls back to sample_id if no id role exists.
